File: 2025-02-February-GFG-POTD/February 06 - Construct Tree from Inorder & Preorder/Solution.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Node {
    int data;
    Node left, right;

    Node(int key) {
        data = key;
        left = right = null;
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            Node root = null;

            // Read inorder array
            String[] inorderStr = (br.readLine()).trim().split(" ");
            int inorder[] = new int[inorderStr.length];
            for (int i = 0; i < inorderStr.length; i++) {
                inorder[i] = Integer.parseInt(inorderStr[i]);
            }

            // Read preorder array
            String[] preorderStr = (br.readLine()).trim().split(" ");
            int preorder[] = new int[preorderStr.length];
            for (int i = 0; i < preorderStr.length; i++) {
                preorder[i] = Integer.parseInt(preorderStr[i]);
            }

            Solution ob = new Solution();
            root = ob.buildTree(inorder, preorder);
            postOrder(root);
            System.out.println();
        }
    }

    // Function to print postorder traversal of the tree
    public static void postOrder(Node root) {
        if (root == null) return;

        postOrder(root.left);
        postOrder(root.right);
        System.out.print(root.data + " ");
    }
}

// } Driver Code Ends


/*
class Node {
    int data;
    Node left, right;

    Node(int key) {
        data = key;
        left = right = null;
    }
}
*/

class Solution {
    static int preIndex = 0;
    
    public static Node buildTree(int inorder[], int preorder[]) {
        preIndex = 0; // Reset index for multiple test cases
        Map<Integer, Integer> inorderMap = new HashMap<>();

        // Store the indices of inorder elements for quick lookup
        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }

        return constructTree(inorder, preorder, 0, inorder.length - 1, inorderMap);
    }

    private static Node constructTree(int inorder[], int preorder[], int inStart, int inEnd, Map<Integer, Integer> inorderMap) {
        if (inStart > inEnd) {
            return null;
        }

        // Pick the current node from Preorder traversal
        int rootValue = preorder[preIndex++];
        Node root = new Node(rootValue);

        // Find the root index in inorder traversal
        int inIndex = inorderMap.get(rootValue);

        // Recursively build the left and right subtrees
        root.left = constructTree(inorder, preorder, inStart, inIndex - 1, inorderMap);
        root.right = constructTree(inorder, preorder, inIndex + 1, inEnd, inorderMap);

        return root;
    }

}

File: 2025-02-February-GFG-POTD/February 06 - Construct Tree from Inorder & Preorder/Solution.py
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int key) {
        data = key;
        left = right = NULL;
    }
};

class Solution {
public:
    int preIndex = 0;
    unordered_map<int, int> inorderMap;

    Node* buildTree(vector<int>& inorder, vector<int>& preorder) {
        preIndex = 0;
        inorderMap.clear();

        // Store indices of inorder elements for quick lookup
        for (int i = 0; i < inorder.size(); i++) {
            inorderMap[inorder[i]] = i;
        }

        return constructTree(inorder, preorder, 0, inorder.size() - 1);
    }

private:
    Node* constructTree(vector<int>& inorder, vector<int>& preorder, int inStart, int inEnd) {
        if (inStart > inEnd) {
            return NULL;
        }

        // Pick the current node from Preorder traversal
        int rootValue = preorder[preIndex++];
        Node* root = new Node(rootValue);

        // Find the root index in inorder traversal
        int inIndex = inorderMap[rootValue];

        // Recursively build the left and right subtrees
        root->left = constructTree(inorder, preorder, inStart, inIndex - 1);
        root->right = constructTree(inorder, preorder, inIndex + 1, inEnd);

        return root;
    }
};

// Function to print postorder traversal of the tree
void postOrder(Node* root) {
    if (root == NULL) return;

    postOrder(root->left);
    postOrder(root->right);
    cout << root->data << " ";
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> inorder(n), preorder(n);

        for (int i = 0; i < n; i++) cin >> inorder[i];
        for (int i = 0; i < n; i++) cin >> preorder[i];

        Solution obj;
        Node* root = obj.buildTree(inorder, preorder);
        postOrder(root);
        cout << endl;
    }
    return 0;
}
